scripts/commands/checkLinks.ts

```diff
@@ -0,0 +1,132 @@
+// This code is a Qiskit project.
+//
+// (C) Copyright IBM 2023.
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+import { globby } from "globby";
+const { existsSync, readFileSync } = require('fs');
+const path = require('node:path')
+const markdownLinkExtractor = require('markdown-link-extractor');
+
+const DOCS_ROOT = "./docs"
+const CONTENT_FILE_EXTENSIONS = [".md", ".mdx", ".ipynb"]
+
+class Link {
+  value: string
+  anchor: string
+  origin: string
+  isExternal: boolean
+
+  constructor(linkString: string, origin: string) {
+    /* 
+     * linkString: Link as it appears in source file
+     *     origin: Path to source file containing link
+     */
+
+    const splitLink = linkString.split('#', 1)
+    this.value = splitLink[0]
+    this.anchor = (splitLink.length > 1) ? `#${splitLink[1]}` : ''
+    this.origin = origin
+    this.isExternal = linkString.startsWith("http")
+  }
+
+  resolve(): string[] {
+    /*
+     * Return list of possible paths link could resolve to
+     */
+    if ( this.isExternal ) { return [ this.value ] }
+    if ( this.value === '' ) { return [ this.origin ] }
+    if ( this.value.startsWith("/images") ) {
+      return [ path.join("public/", this.value) ]
+    }
+
+    let baseFilePath
+    if (this.value.startsWith('/')) {
+      // Path is relative to DOCS_ROOT
+      baseFilePath = path.join(DOCS_ROOT, this.value)
+    } else {
+      // Path is relative to origin file
+      baseFilePath = path.join(
+        path.dirname(this.origin),
+        this.value
+      )
+    }
+    // Remove trailing '/' from path.join
+    baseFilePath = baseFilePath.replace(/\/$/gm, '')
+
+    // File may have one of many extensions (.md, .ipynb etc.), and/or be
+    // directory with an index file (e.g. `docs/build` should resolve to
+    // `docs/build/index.mdx`). We return a list of possible filenames.
+    let possibleFilePaths = []
+    for (let index of ['', '/index']) {
+      for (let extension of CONTENT_FILE_EXTENSIONS) {
+        possibleFilePaths.push(
+            baseFilePath + index + extension
+        )
+      }
+    }
+    return possibleFilePaths
+  }
+
+  check(): boolean {
+    /*
+     * True if link points to existing file, otherwise false
+     */
+    if (this.isExternal) {
+      // External link checking not supported yet
+      return true
+    }
+
+    for (let filePath of this.resolve()) {
+      if (existsSync(filePath)) {
+        return true
+      }
+    }
+
+    console.log(`❌ ${this.origin}: Could not find link '${this.value}'`)
+    return false
```

Comment:
Instead of logging directly in this function, it might be a cleaner user interface to collect all the failing links in the `main` function or `checkLinksInFile`, and then bundle them together as a single console error, where you list all the failures.

This is a lot of repeated warnings right now, which can be noisy.

See an example: https://github.com/Qiskit/documentation/blob/a1a72fe18a8fed759b733965ef0fc6440e303ebc/scripts/commands/checkMetadata.ts#L55-L65

You'd have this function return something like structured data or a string representing what the failure is. Return `undefined` if it's okay. Return type is `string?`.